File: NestCash/backend/app/models/forum_models.py
# app/models/forum_models.py
from datetime import datetime
from typing import List, Optional, Dict
from pydantic import BaseModel, Field
from beanie import Document, PydanticObjectId
from enum import Enum

class PrivacyLevel(str, Enum):
    PUBLIC = "public"          # Mindenki láthatja
    FRIENDS = "friends"        # Csak követők láthatják
    PRIVATE = "private"        # Csak saját maga láthatja

class PostCategory(str, Enum):
    GENERAL = "general"
    BUDGETING = "budgeting"
    INVESTING = "investing"
    SAVINGS = "savings"
    CAREER = "career"
    EXPENSES = "expenses"
    TIPS = "tips"
    QUESTIONS = "questions"

class FeedType(str, Enum):
    ALL = "all"               # Összes poszt (publikus + követettek)
    FOLLOWING = "following"   # Csak követettek posztjai
    MY_POSTS = "my_posts"     # Csak saját posztok

class SortBy(str, Enum):
    NEWEST = "newest"         # Legújabb először
    POPULAR = "popular"       # Legnépszerűbb (like-ok alapján)
    MOST_COMMENTED = "most_commented"  # Legtöbb komment

class NotificationType(str, Enum):
    LIKE = "like"
    COMMENT = "comment"
    FOLLOW = "follow"

# === FORUM POST DOCUMENT ===
class ForumPostDocument(Document):
    user_id: PydanticObjectId
    username: str  # Denormalizált adat a gyorsabb lekérdezésekért
    title: str
    content: str
    category: PostCategory
    privacy_level: PrivacyLevel = Field(default=PrivacyLevel.PUBLIC)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    like_count: int = Field(default=0)
    comment_count: int = Field(default=0)
    
    class Settings:
        name = "forum_posts"

# === COMMENT DOCUMENT ===
class CommentDocument(Document):
    post_id: PydanticObjectId
    user_id: PydanticObjectId
    username: str  # Denormalizált adat
    content: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "forum_comments"

# === LIKE DOCUMENT ===
class LikeDocument(Document):
    post_id: PydanticObjectId
    user_id: PydanticObjectId
    username: str  # Denormalizált adat
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "forum_likes"

# === FOLLOW DOCUMENT ===
class FollowDocument(Document):
    follower_id: PydanticObjectId  # Ki követ
    following_id: PydanticObjectId # Kit követ
    follower_username: str
    following_username: str
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "forum_follows"

# === NOTIFICATION DOCUMENT ===
class NotificationDocument(Document):
    user_id: PydanticObjectId  # Kinek szól az értesítés
    from_user_id: PydanticObjectId  # Ki váltotta ki
    from_username: str
    type: NotificationType
    post_id: Optional[PydanticObjectId] = None  # Ha poszt-related
    message: str
    is_read: bool = Field(default=False)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "forum_notifications"

# === USER FORUM SETTINGS DOCUMENT ===
class UserForumSettingsDocument(Document):
    user_id: PydanticObjectId
    default_privacy_level: PrivacyLevel = Field(default=PrivacyLevel.PUBLIC)
    notifications_enabled: Dict[str, bool] = Field(default={
        "like": True,
        "comment": True,
        "follow": True
    })
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "forum_user_settings"

# === RESPONSE SCHEMAS ===
class PostCreate(BaseModel):
    title: str = Field(..., min_length=3, max_length=200)
    content: str = Field(..., min_length=10, max_length=5000)
    category: PostCategory
    privacy_level: PrivacyLevel = Field(default=PrivacyLevel.PUBLIC)

class PostUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=3, max_length=200)
    content: Optional[str] = Field(None, min_length=10, max_length=5000)
    category: Optional[PostCategory] = None
    privacy_level: Optional[PrivacyLevel] = None

class PostRead(BaseModel):
    id: str
    user_id: str
    username: str
    title: str
    content: str
    category: PostCategory
    privacy_level: PrivacyLevel
    created_at: datetime
    updated_at: datetime
    like_count: int
    comment_count: int
    is_liked_by_me: bool = Field(default=False)  # Csak válaszban
    is_my_post: bool = Field(default=False)      # Csak válaszban

class PostListResponse(BaseModel):
    posts: List[PostRead]
    total_count: int
    skip: int
    limit: int

class CommentCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=1000)

class CommentRead(BaseModel):
    id: str
    post_id: str
    user_id: str
    username: str
    content: str
    created_at: datetime
    updated_at: datetime
    is_my_comment: bool = Field(default=False)

class CommentListResponse(BaseModel):
    comments: List[CommentRead]
    total_count: int

class UserSearch(BaseModel):
    id: str
    username: str
    is_following: bool = Field(default=False)
    is_followed_by: bool = Field(default=False)

class UserSearchResponse(BaseModel):
    users: List[UserSearch]
    total_count: int

class FollowRead(BaseModel):
    follower_id: str
    following_id: str
    follower_username: str
    following_username: str
    created_at: datetime

class FollowListResponse(BaseModel):
    follows: List[FollowRead]
    total_count: int

class NotificationRead(BaseModel):
    id: str
    from_user_id: str
    from_username: str
    type: NotificationType
    post_id: Optional[str] = None
    message: str
    is_read: bool
    created_at: datetime

class NotificationListResponse(BaseModel):
    notifications: List[NotificationRead]
    unread_count: int
    total_count: int

class ForumSettingsRead(BaseModel):
    default_privacy_level: PrivacyLevel
    notifications_enabled: Dict[str, bool]

class ForumSettingsUpdate(BaseModel):
    default_privacy_level: Optional[PrivacyLevel] = None
    notifications_enabled: Optional[Dict[str, bool]] = None

class ForumStatsResponse(BaseModel):
    my_posts_count: int
    my_likes_received: int
    followers_count: int
    following_count: int

File: NestCash/frontend/lib/models/forum_models.dart
// lib/models/forum_models.dart

class ForumPost {
  final String id;
  final String userId;
  final String username;
  final String title;
  final String content;
  final String category;
  final String privacyLevel;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int likeCount;
  final int commentCount;
  final bool isLikedByMe;
  final bool isMyPost;

  ForumPost({
    required this.id,
    required this.userId,
    required this.username,
    required this.title,
    required this.content,
    required this.category,
    required this.privacyLevel,
    required this.createdAt,
    required this.updatedAt,
    required this.likeCount,
    required this.commentCount,
    required this.isLikedByMe,
    required this.isMyPost,
  });

  factory ForumPost.fromJson(Map<String, dynamic> json) {
    return ForumPost(
      id: json['id'],
      userId: json['user_id'],
      username: json['username'],
      title: json['title'],
      content: json['content'],
      category: json['category'],
      privacyLevel: json['privacy_level'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      likeCount: json['like_count'],
      commentCount: json['comment_count'],
      isLikedByMe: json['is_liked_by_me'] ?? false,
      isMyPost: json['is_my_post'] ?? false,
    );
  }

  String get categoryDisplayName {
    switch (category) {
      case 'general':
        return 'Általános';
      case 'budgeting':
        return 'Költségvetés';
      case 'investing':
        return 'Befektetés';
      case 'savings':
        return 'Megtakarítás';
      case 'career':
        return 'Karrier';
      case 'expenses':
        return 'Kiadások';
      case 'tips':
        return 'Tippek';
      case 'questions':
        return 'Kérdések';
      default:
        return category;
    }
  }
}

class ForumComment {
  final String id;
  final String postId;
  final String userId;
  final String username;
  final String content;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isMyComment;

  ForumComment({
    required this.id,
    required this.postId,
    required this.userId,
    required this.username,
    required this.content,
    required this.createdAt,
    required this.updatedAt,
    required this.isMyComment,
  });

  factory ForumComment.fromJson(Map<String, dynamic> json) {
    return ForumComment(
      id: json['id'],
      postId: json['post_id'],
      userId: json['user_id'],
      username: json['username'],
      content: json['content'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      isMyComment: json['is_my_comment'] ?? false,
    );
  }
}

class ForumUser {
  final String id;
  final String username;
  final bool isFollowing;
  final bool isFollowedBy;

  ForumUser({
    required this.id,
    required this.username,
    required this.isFollowing,
    required this.isFollowedBy,
  });

  factory ForumUser.fromJson(Map<String, dynamic> json) {
    return ForumUser(
      id: json['id'],
      username: json['username'],
      isFollowing: json['is_following'] ?? false,
      isFollowedBy: json['is_followed_by'] ?? false,
    );
  }
}

class ForumNotification {
  final String id;
  final String fromUserId;
  final String fromUsername;
  final String type;
  final String? postId;
  final String message;
  final bool isRead;
  final DateTime createdAt;

  ForumNotification({
    required this.id,
    required this.fromUserId,
    required this.fromUsername,
    required this.type,
    this.postId,
    required this.message,
    required this.isRead,
    required this.createdAt,
  });

  factory ForumNotification.fromJson(Map<String, dynamic> json) {
    return ForumNotification(
      id: json['id'],
      fromUserId: json['from_user_id'],
      fromUsername: json['from_username'],
      type: json['type'],
      postId: json['post_id'],
      message: json['message'],
      isRead: json['is_read'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  String get typeDisplayName {
    switch (type) {
      case 'like':
        return 'Kedvelés';
      case 'comment':
        return 'Komment';
      case 'follow':
        return 'Követés';
      default:
        return type;
    }
  }
}

class ForumStats {
  final int myPostsCount;
  final int myLikesReceived;
  final int followersCount;
  final int followingCount;

  ForumStats({
    required this.myPostsCount,
    required this.myLikesReceived,
    required this.followersCount,
    required this.followingCount,
  });

  factory ForumStats.fromJson(Map<String, dynamic> json) {
    return ForumStats(
      myPostsCount: json['my_posts_count'],
      myLikesReceived: json['my_likes_received'],
      followersCount: json['followers_count'],
      followingCount: json['following_count'],
    );
  }
}

enum PostCategory {
  general,
  budgeting,
  investing,
  savings,
  career,
  expenses,
  tips,
  questions,
}

extension PostCategoryExtension on PostCategory {
  String get value {
    switch (this) {
      case PostCategory.general:
        return 'general';
      case PostCategory.budgeting:
        return 'budgeting';
      case PostCategory.investing:
        return 'investing';
      case PostCategory.savings:
        return 'savings';
      case PostCategory.career:
        return 'career';
      case PostCategory.expenses:
        return 'expenses';
      case PostCategory.tips:
        return 'tips';
      case PostCategory.questions:
        return 'questions';
    }
  }

  String get displayName {
    switch (this) {
      case PostCategory.general:
        return 'Általános';
      case PostCategory.budgeting:
        return 'Költségvetés';
      case PostCategory.investing:
        return 'Befektetés';
      case PostCategory.savings:
        return 'Megtakarítás';
      case PostCategory.career:
        return 'Karrier';
      case PostCategory.expenses:
        return 'Kiadások';
      case PostCategory.tips:
        return 'Tippek';
      case PostCategory.questions:
        return 'Kérdések';
    }
  }
}

enum FeedType {
  all,
  following,
  myPosts,
}

extension FeedTypeExtension on FeedType {
  String get value {
    switch (this) {
      case FeedType.all:
        return 'all';
      case FeedType.following:
        return 'following';
      case FeedType.myPosts:
        return 'my_posts';
    }
  }

  String get displayName {
    switch (this) {
      case FeedType.all:
        return 'Minden poszt';
      case FeedType.following:
        return 'Követettek';
      case FeedType.myPosts:
        return 'Saját posztok';
    }
  }
}

enum SortBy {
  newest,
  popular,
  mostCommented,
}

extension SortByExtension on SortBy {
  String get value {
    switch (this) {
      case SortBy.newest:
        return 'newest';
      case SortBy.popular:
        return 'popular';
      case SortBy.mostCommented:
        return 'most_commented';
    }
  }

  String get displayName {
    switch (this) {
      case SortBy.newest:
        return 'Legújabb';
      case SortBy.popular:
        return 'Legnépszerűbb';
      case SortBy.mostCommented:
        return 'Legtöbb komment';
    }
  }
}